include/bl_storage.h

```diff
@@ -7,15 +7,75 @@
 #ifndef BL_STORAGE_H_
 #define BL_STORAGE_H_
 
-#include <zephyr/types.h>
 #include <string.h>
+#include <zephyr/types.h>
+#include <drivers/nrfx_common.h>
+#include <nrfx_nvmc.h>
+#include <errno.h>
+#include <pm_config.h>
+/* In case this gets included in an image that is build without a bootloader
+ * the provision area is not defined, so we have to point to the OTP manually
+ * NCSDK-18032 is to prevent this workaround in the future
+ */
+#ifndef PM_PROVISION_ADDRESS
+#define PM_PROVISION_ADDRESS (&NRF_UICR->OTP)
+#endif
+
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-
 #define EHASHFF 113 /* A hash contains too many 0xFs. */
+#define EREADLCS 114 /* LCS field of OTP is in an invalid state */
+#define EINVALIDLCS 115 /* Invalid LCS*/
+
+/* We truncate the 32 byte sha256 down to 16 bytes before storing it */
+#define SB_PUBLIC_KEY_HASH_LEN 16
+
+/** Storage for the PRoT Security Lifecycle state, that consists of 4 states:
+ *  - Device assembly and test
+ *  - PRoT Provisioning
+ *  - Secured
+ *  - Decommissioned
+ *  These states are transitioned top down during the life time of a device.
+ *  The Device assembly and test state is implicitly defined by checking if
+ *  the provisioning state wasn't entered yet.
+ *  This works as ASSEMBLY implies the OTP to be erased.
+ */
+struct life_cycle_state_data {
+	uint16_t provisioning;
+	uint16_t secure;
+	/* Pad to end the alignment at a 4-byte boundary as the UICR->OTP
+	 * only supports 4-byte reads. We place the reserved padding in
+	 * the middle of the struct in case we ever need to support
+	 * another state.
+	 */
+	uint16_t reserved_for_padding;
+	uint16_t decommissioned;
+};
+
+/** The first data structure in the bootloader storage. It has unknown length
+ *  since 'key_data' is repeated. This data structure is immediately followed by
+ *  struct counter_collection.
+ */
+struct bl_storage_data {
+	/* NB: When placed in OTP, reads must be 4 bytes and 4 byte aligned */
+	struct life_cycle_state_data lcs;
+	uint8_t implementation_id[32];
+	uint32_t s0_address;
+	uint32_t s1_address;
+	uint32_t num_public_keys; /* Number of entries in 'key_data' list. */
+	struct {
+		uint32_t valid;
+		uint8_t hash[SB_PUBLIC_KEY_HASH_LEN];
+	} key_data[1];
+};
+
+#define BL_STORAGE ((struct bl_storage_data *)(PM_PROVISION_ADDRESS))
```

Comment:
```suggestion
#define BL_STORAGE ((const struct bl_storage_data *)(PM_PROVISION_ADDRESS))
```
?

Reply:
why?

Reply:
Hey, that worked great :)

Not sure how often variables are accidentally written to (I very rarely see this warning), but I haven't seen any sign of annoying const propogation, so this looks safe :)

/home/sebo/ncs/nrf/include/bl_storage.h: In function 'read_implementation_id_from_otp':
/home/sebo/ncs/nrf/include/bl_storage.h:214:42: error: assignment of read-only location '16744704->implementation_id[0]'
  214 |         BL_STORAGE->implementation_id[0] = 3;

Reply:
Maybe "const volatile" is the most appropriate?

To reflect the fact that this is not normal const, it can change when we update the LCS, counters, etc.

Reply:
// read from const struct
// write to flash
// read same value from const struct.

This might fail without volatile I think.